sqlserver/datadog_checks/sqlserver/sqlserver.py

```diff
@@ -754,6 +785,12 @@ def check(self, _):
             if self._config.autodiscovery and self._config.autodiscovery_db_service_check:
                 self._check_database_conns()
             self._send_database_instance_metadata()
+            if (
+                self._last_schemas_collect_time is None
+                or time.time() - self._last_schemas_collect_time > self._config.schemas_collection_interval
+            ):
+                self._schemas.collect_schemas_data()
+                self._last_schemas_collect_time = time.time()
```

Comment:
I think `_last_schemas_collect_time` should not be a agent check property but be a private property in `_schema`. This way the control flow can be managed within the `class Schema`

Reply:
> We might want to reverse these lines so that we set schemas as collected first in case the collect_schemas_data throws an exception. Here is an example of why this could have a negative impact:

I think this code should just be kicked off in a separate thread && then (like @lu-zhengda already suggested) the control flow can be entirely managed in the schema class.

Reply:
I saw check running forever (like more than 10 mins ) in agent-dev. 

We had some discussions with Zhengda where we thought of running it in the same thread this is cause he spotted that threads have a negative impact on performance. But I see your point too that we now loose control over execution time of schema collection(DBMAsyncJob are getting killed for inactivity). What if we add a time limit for schema collection ? Say 3 minutes ? And then stop it and report to the backend ? Would 3 mins mess up peoples metrics ?

@lu-zhengda  - is the impact on performance is that bad (I can try to measure it too but some order of magnitude number ?) It is nice to be able to leverage DBMAsyncJob code though.

Reply:
> We might want to reverse these lines so that we set schemas as collected first in case the `collect_schemas_data` throws an exception. Here is an example of why this could have a negative impact:
> 
> 1. Agent starts up, begins collecting schema data.
> 2. An exception is thrown in the `submit` function and the check fails
> 3. 15 seconds later, the Agent runs the check again and `_last_schemas_collect_time` is None
> 4. steps 2-3 are repeated infinitely
> 
> In this case, we accidentally run and expensive set of queries every 15 seconds instead of every 1200 seconds.
> 
> However, this also means that if schema collection fails, you miss data for 20 minutes. A more complicated alternative is to set a different rate limit time for failures vs. successes. So successes are rate limited every 20 minutes and failures are rate limited every 5 minutes.

I looked at the DBMAsyncJob - cant see a logic there that tries to deal with a happy and unhappy case rate. I think not to overcomplicate for now - make sure that its not executed more often than X minutes but if it fails then just wait.

Reply:
Is it possible to introduce a time limit for the safety reasons, and start where we left in the next check run?

Reply:
this is one of the 2 optimisations that I wanted to have as a follow up if we decide to do so.
We can also decide that it has to be in this version. I am open to that. I actually have some draft code.